Source/Core/Core/FifoPlayer/FifoPlayer.cpp

```diff
@@ -414,38 +392,33 @@ void FifoPlayer::WriteFrame(const FifoFrameInfo& frame, const AnalyzedFrameInfo&
   m_ElapsedCycles = 0;
   m_FrameFifoSize = static_cast<u32>(frame.fifoData.size());
 
-  // Determine start and end objects
-  const u32 num_objects = (u32)(info.objects.size());
-  const u32 draw_start = std::min(num_objects, m_ObjectRangeStart);
-  const u32 draw_end = std::min(num_objects - 1, m_ObjectRangeEnd);
-
   u32 memory_update = 0;
+  u32 object_num = 0;
 
-  // Skip memory updates during frame if true
+  // Skip all memory updates if early memory updates are enabled, as we already wrote them
   if (m_EarlyMemoryUpdates)
   {
     memory_update = (u32)(frame.memoryUpdates.size());
   }
 
-  for (u32 object_num = 0; object_num < num_objects; object_num++)
+  for (const FramePart& part : info.parts)
   {
-    const auto& object = info.objects[object_num];
-    if (object_num < draw_start || object_num > draw_end)
+    bool show_part;
+
+    if (part.m_type == FramePartType::PrimitiveData)
     {
-      // If an object is out of range, only apply the commands, not the primitive data.
-      // We need to still apply the commands because commands from earlier objects still apply to
-      // later ones (and generally games will not reconfigure everything for each object).
-      WriteFramePart(object.m_start, object.m_start + object.m_primitive_offset, memory_update,
-                     frame);
+      show_part = m_ObjectRangeStart <= object_num && object_num <= m_ObjectRangeEnd;
+      object_num++;
     }
     else
     {
-      ASSERT(object_num == draw_start);
-      const auto& last_object = info.objects[draw_end];
-      object_num = draw_end;
-      WriteFramePart(object.m_start, last_object.m_start + last_object.m_size, memory_update,
-                     frame);
+      // We always include commands and EFB copies, as commands from earlier objects still apply to
+      // later ones (games generally do not reconfigure everything for each object)
+      show_part = true;
     }
+
+    if (show_part)
+      WriteFramePart(part, &memory_update, frame);
   }
```

Comment:
The new code, although a lot simpler, results in [fifoci differences](https://fifo.ci/version/6fd946c70f69f8d69ab52f80f133c4d54ab15b57/) (most noticeable with [ntfu-reflections](https://fifo.ci/compare/7043712-7041897/)).  These are due to different timings caused by using several calls to `WriteFramePart` instead of one big call to `WriteFramePart` for the main object range (particularly the common case where all objects are enabled).

I've switched back to one big `WriteFramePart` call for now, but it might be better to just accept the timing differences.

Reply:
I switched back to the newer code again.